azuredevops/internal/service/feed/resource_feed.go

```diff
@@ -0,0 +1,136 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/v7/feed"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/client"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils"
+)
+
+func ResourceFeed() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceFeedCreate,
+		Read:   resourceFeedRead,
+		Update: resourceFeedUpdate,
+		Delete: resourceFeedDelete,
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				ValidateFunc: validation.StringIsNotWhiteSpace,
+				Required:     true,
+			},
+			"project_id": {
+				Type:         schema.TypeString,
+				ValidateFunc: validation.StringIsNotWhiteSpace,
+				Optional:     true,
+			},
+		},
+	}
+}
+
+func resourceFeedCreate(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+	name := d.Get("name").(string)
+	projectId := d.Get("project_id").(string)
+
+	createFeed := feed.Feed{
+		Name: &name,
+	}
+
+	createdFeed, err := clients.FeedClient.CreateFeed(clients.Ctx, feed.CreateFeedArgs{
+		Feed:    &createFeed,
+		Project: &projectId,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	d.SetId((*createdFeed).Id.String())
+
+	return resourceFeedRead(d, m)
+}
+
+func resourceFeedRead(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+
+	name := d.Get("name").(string)
+	projectId := d.Get("project_id").(string)
+
+	getFeed, err := clients.FeedClient.GetFeed(clients.Ctx, feed.GetFeedArgs{
+		FeedId:  &name,
+		Project: &projectId,
+	})
+
+	if err != nil {
+		if utils.ResponseWasNotFound(err) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error reading feed during read: %+v", err)
+	}
+
+	if getFeed != nil {
+		d.SetId((*getFeed).Id.String())
+		d.Set("name", (*getFeed).Name)
+		project := (*getFeed).Project
+		if project != nil {
+			d.Set("project_id", (*project).Id.String())
+		}
+	}
+
+	return nil
+}
+
+func resourceFeedUpdate(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+	name := d.Get("name").(string)
+	projectId := d.Get("project_id").(string)
+
+	updateFeed := &feed.FeedUpdate{}
+
+	updatedFeed, err := clients.FeedClient.UpdateFeed(clients.Ctx, feed.UpdateFeedArgs{
+		Feed:    updateFeed,
+		FeedId:  &name,
+		Project: &projectId,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	d.SetId((*updatedFeed).Id.String())
+
+	return resourceFeedRead(d, m)
+}
+
+func resourceFeedDelete(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+	name := d.Get("name").(string)
+	project_id := d.Get("project_id").(string)
+
+	err := clients.FeedClient.DeleteFeed(clients.Ctx, feed.DeleteFeedArgs{
+		FeedId:  &name,
+		Project: &project_id,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	err = clients.FeedClient.PermanentDeleteFeed(clients.Ctx, feed.PermanentDeleteFeedArgs{
```

Comment:
Considering that the feed can be restored, is it possible to change permanent deletion to soft deletion?

Reply:
What about recreations ? Cause if for some reason there will be recreate happening, this will fail, as previously Feed was not removed, just moved to the bin. We would have to consider how to handle "lifecycle" with softdelete, and i could see there were already some discussions in the Issue section

Reply:
The feed can be restored from the recycle bin. I would suggest adding a flag if we need to permanently delete the feed and the creation can check the recycle bin if and restore the feed if needed.
API:
https://learn.microsoft.com/en-us/rest/api/azure/devops/artifacts/feed-recycle-bin/restore-deleted-feed?view=azure-devops-rest-7.1

Reply:
Sure, i will add that in a moment

Reply:
Hello @xuzhang3, I have added support for soft delete, please verify it works for you

Reply:
I have also introduced `restored` computed value, which is set depending on Feed being created or restored